url29ae058134_yuriy_chumak_jatha_tgz-pJ8-org_jatha_LispJ8/src/org/jatha/read/LispParser.java

```diff
@@ -0,0 +1,1065 @@
+	static final int READING_BACKQUOTED_LIST   = 5;
+
+	private      int BackQuoteLevel            = 0;
+	private      PushbackReader  inputReader;
+
+	private int f_caseSensitivity = UPCASE;  // default LISP behavior.
+
+	private static LispParser f_myParser = null;
+
+
+	// required by tokenToLispValue() to use "intern" and "keyword" functions
+	private Lisp f_lisp = null;
+
+	public LispParser(Lisp lisp, Reader inStream)
+	{
+		this(lisp, inStream, UPCASE);
+	}
+
+	public LispParser(Lisp lisp, String inString)
+	{
+		this(lisp, new StringReader(inString), UPCASE);
+	}
+
+
+	/**
+	 * Allows you to create a parser that handles
+	 * input case conversion as you like.
+	 * Default is UPCASE.  Other values are DOWNCASE and PRESERVE.
+	 * @param inStream
+	 */
+	public LispParser(Lisp lisp, Reader inStream, int caseSensitivity)
+	{
+		f_lisp = lisp;
+
+		if (inStream instanceof PushbackReader)
+			inputReader = (PushbackReader)inStream;
+		else
+			inputReader   = new PushbackReader(inStream);
+		setCaseSensitivity(caseSensitivity);
+	}
+
+	/**
+	 * Allows you to create a parser that handles
+	 * input case conversion as you like.
+	 * Default is UPCASE.  Other values are DOWNCASE and PRESERVE.
+	 */
+	public LispParser(Lisp lisp, String inString, int caseSensitivity)
+	{
+		this(lisp, new StringReader(inString));
+		setCaseSensitivity(caseSensitivity);
+	}
+
+  /**
+   * Retrieves the current case-sensitivity of the parser.
+   * It can be eiher LispParser.UPCASE, LispParser.DOWNCASE
+   * or LispParser.PRESERVE
+   * @return UPCASE, DOWNCASE or PRESERVE
+   */
+  public int getCaseSensitivity()
+  {
+    return f_caseSensitivity;
+  }
+
+  /**
+   * Sets the current case-sensitivity of the parser.
+   * It can be eiher LispParser.UPCASE, LispParser.DOWNCASE
+   * or LispParser.PRESERVE
+   */
+  public void setCaseSensitivity(int caseSensitivity)
+  {
+    f_caseSensitivity = caseSensitivity;
+  }
+
+
+  /**
+   * Gets the current reader to be parsed.
+   */
+  public PushbackReader getInputReader()
+  {
+    return inputReader;
+  }
+
+
+  /**
+   * Sets the input reader for the Parser.
+   */
+  public void setInputReader(PushbackReader inputReader)
+  {
+    this.inputReader = inputReader;
+  }
+
+
+  /**
+   * Sets the input string for the parser.  This is the
+   * String to parse.
+   */
+  public void setInputString(String s)
+  {
+    this.inputReader = new PushbackReader(new StringReader(s));
+  }
+
+
+	/**
+	 * Parse() assumes that there is only one expression in the input
+	 * string or file.  If you need to read multiple items from a
+	 * string or file, use the read() function.
+	 * Parse just calls read right now.
+	 *
+	 * @see #read
+	 */
+	public LispValue parse()
+			throws EOFException
+	{
+		return read();
+	}
+
+	/**
+	 * Reads one s-expression from the input stream (a string or file).
+	 * Throws an EOFxception when EOF is reached.
+	 * Call this method repeatedly to do read-eval-print on a file.
+	 */
+	public LispValue read()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 31. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.